clientapi/routing/register.go

```diff
@@ -496,11 +497,20 @@ func Register(
 		r.Username = strconv.FormatInt(id, 10)
 	}
 
+	// Is this an appservice registration? It will be if the access
+	// token is supplied
+	accessToken, accessTokenErr := auth.ExtractAccessToken(req)
+
 	// Squash username to all lowercase letters
 	r.Username = strings.ToLower(r.Username)
-
-	if resErr = validateUsername(r.Username); resErr != nil {
-		return *resErr
+	if r.Auth.Type == authtypes.LoginTypeApplicationService || accessTokenErr == nil {
```

Comment:
I'd prefer it if we only look at the access token if the login type is set to AS?

Reply:
It turns out that this isn't always the case. It's another one of those "spec is vague" things but Synapse also treats all calls to `/register` with an `access_token` as if they are coming from an AS.

Reply:
Decided with @Half-Shot to instead enforce `authtypes.LoginTypeApplicationService` only.